File: sdks/java/connect-sdk/src/main/java/com/mobiscroll/connect/resources/Events.java
package com.mobiscroll.connect.resources;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.mobiscroll.connect.ApiClient;
import com.mobiscroll.connect.Provider;
import com.mobiscroll.connect.exceptions.MobiscrollConnectException;
import com.mobiscroll.connect.internal.JsonMapperHolder;
import com.mobiscroll.connect.internal.QueryStringBuilder;
import com.mobiscroll.connect.models.CalendarEvent;
import com.mobiscroll.connect.models.EventCreateData;
import com.mobiscroll.connect.models.EventDeleteParams;
import com.mobiscroll.connect.models.EventListParams;
import com.mobiscroll.connect.models.EventUpdateData;
import com.mobiscroll.connect.models.EventsListResponse;

/** Events resource: list, create, update, delete. */
public final class Events {

    private final ApiClient api;

    public Events(ApiClient api) {
        this.api = api;
    }

    /**
     * List events across the supplied calendars (or all connected calendars if none specified).
     * The {@code calendarIds} map is JSON-encoded into a single query parameter, matching the
     * wire format used by the other SDKs (e.g. {@code calendarIds={"google":["primary"]}}).
     */
    public EventsListResponse list(EventListParams params) {
        QueryStringBuilder qs = new QueryStringBuilder();
        if (params != null) {
            Map<Provider, List<String>> ids = params.getCalendarIds();
            if (ids != null && !ids.isEmpty()) {
                Map<String, List<String>> wire = new LinkedHashMap<>();
                for (Map.Entry<Provider, List<String>> e : ids.entrySet()) {
                    wire.put(e.getKey().wireValue(), e.getValue());
                }
                try {
                    qs.add("calendarIds", JsonMapperHolder.get().writeValueAsString(wire));
                } catch (JsonProcessingException e) {
                    throw new MobiscrollConnectException("Failed to encode calendarIds", e);
                }
            }
            qs.add("start", params.getStart())
              .add("end", params.getEnd())
              .add("pageSize", params.getPageSize())
              .add("nextPageToken", params.getNextPageToken())
              .add("singleEvents", params.getSingleEvents());
        }
        return api.get("/events", qs.isEmpty() ? null : qs.build(),
                new TypeReference<EventsListResponse>() {});
    }

    /** Create an event. Uses singular {@code /event} path (matches server contract). */
    public CalendarEvent create(EventCreateData data) {
        return api.post("/event", data, new TypeReference<CalendarEvent>() {});
    }

    /** Update an event. Uses singular {@code /event} path (matches server contract). */
    public CalendarEvent update(EventUpdateData data) {
        return api.put("/event", data, new TypeReference<CalendarEvent>() {});
    }

    /** Delete an event. Uses singular {@code /event} path (matches server contract). */
    public void delete(EventDeleteParams params) {
        QueryStringBuilder qs = new QueryStringBuilder()
                .add("provider", params.getProvider().wireValue())
                .add("calendarId", params.getCalendarId())
                .add("eventId", params.getEventId())
                .add("recurringEventId", params.getRecurringEventId())
                .add("deleteMode", params.getDeleteMode());
        api.delete("/event", qs.build());
    }
}

File: sdks/go/events.go
package mobiscroll

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"

	"github.com/acidb/mobiscroll-connect-sdks/sdks/go/internal/querybuilder"
)

// eventsService groups the events endpoints.
type eventsService struct{ api *apiClient }

// List returns a paginated slice of events across the calendars selected by
// params. Pass nil for default behavior (events for all connected calendars,
// server-default page size). The CalendarIDs map is serialized as a single
// JSON-encoded query parameter (e.g. calendarIds={"google":["primary"]}),
// matching the wire format of the other SDKs.
func (s *eventsService) List(ctx context.Context, params *EventListParams) (*EventsListResponse, error) {
	qs := querybuilder.New()
	if params != nil {
		if len(params.CalendarIDs) > 0 {
			wire := make(map[string][]string, len(params.CalendarIDs))
			for k, v := range params.CalendarIDs {
				wire[string(k)] = v
			}
			encoded, err := json.Marshal(wire)
			if err != nil {
				return nil, fmt.Errorf("encode calendarIds: %w", err)
			}
			qs.Add("calendarIds", string(encoded))
		}
		qs.Add("start", params.Start).
			Add("end", params.End).
			Add("pageSize", params.PageSize).
			Add("nextPageToken", params.NextPageToken).
			Add("singleEvents", params.SingleEvents)
	}
	out := &EventsListResponse{}
	if err := s.api.do(ctx, http.MethodGet, "/events", qs.Encode(), nil, out); err != nil {
		return nil, err
	}
	return out, nil
}

// Create creates a new event and returns the server-assigned representation.
func (s *eventsService) Create(ctx context.Context, data *EventCreateData) (*CalendarEvent, error) {
	if data == nil {
		return nil, errors.New("mobiscroll: EventCreateData must not be nil")
	}
	out := &CalendarEvent{}
	if err := s.api.do(ctx, http.MethodPost, "/event", "", data, out); err != nil {
		return nil, err
	}
	return out, nil
}

// Update updates an existing event. EventID is required; non-zero fields are
// applied as patches.
func (s *eventsService) Update(ctx context.Context, data *EventUpdateData) (*CalendarEvent, error) {
	if data == nil {
		return nil, errors.New("mobiscroll: EventUpdateData must not be nil")
	}
	out := &CalendarEvent{}
	if err := s.api.do(ctx, http.MethodPut, "/event", "", data, out); err != nil {
		return nil, err
	}
	return out, nil
}

// Delete deletes an event (or a slice of a recurring series, when
// RecurringEventID + DeleteMode are set).
func (s *eventsService) Delete(ctx context.Context, params *EventDeleteParams) error {
	if params == nil {
		return errors.New("mobiscroll: EventDeleteParams must not be nil")
	}
	qs := querybuilder.New().
		Add("provider", string(params.Provider)).
		Add("calendarId", params.CalendarID).
		Add("eventId", params.EventID).
		Add("recurringEventId", params.RecurringEventID).
		Add("deleteMode", params.DeleteMode).
		Encode()
	return s.api.do(ctx, http.MethodDelete, "/event", qs, nil, nil)
}
